Zoltan2: Compilation errors with unit tests in Cuda builds with UVM enabled

## Bug Report
@trilinos/zoltan2 
Adding @egboman , can you help with assignment of the issue?

Attempting to compile Zoltan2 unit tests in a Cuda build with UVM enabled errors out with output:
```
/home/ndellin/trilinos/Trilinos-pristine/packages/kokkos/core/src/Kokkos_View.hpp(1334): error: static assertion failed with "Incompatible View copy assignment"
          detected during:
            instantiation of "std::enable_if_t<Kokkos::Impl::ViewMapping<Kokkos::View<DataType, Properties...>::traits, Kokkos::View<RT, RP...>::traits, Kokkos::ViewTraits<DataType, Properties...>::specialize>::is_assignable_data_type, Kokkos::View<DataType, Properties...>> &Kokkos::View<DataType, Properties...>::operator=(const Kokkos::View<RT, RP...> &) [with DataType=std::remove_const_t<const Tpetra::Details::DefaultTypes::global_ordinal_type> *, Properties=<std::conditional_t<true, Kokkos::Cuda::array_layout, Kokkos::Cuda::array_layout>, Kokkos::Device<Kokkos::HostSpace::execution_space, Kokkos::Cuda::memory_space>, std::conditional_t<false, void, Kokkos::Experimental::DefaultViewHooks>>, RT=std::remove_const_t<const Tpetra::Details::DefaultTypes::global_ordinal_type> *, RP=<std::conditional_t<true, Kokkos::Cuda::array_layout, Kokkos::Cuda::array_layout>, Kokkos::Device<Kokkos::HostSpace::execution_space, Kokkos::HostSpace::memory_space>, std::conditional_t<false, void, Kokkos::Experimental::DefaultViewHooks>>]"
/home/ndellin/trilinos/Trilinos-pristine/packages/zoltan2/core/src/input/Zoltan2_TpetraRowMatrixAdapter.hpp(169): here
            instantiation of "void Zoltan2::TpetraRowMatrixAdapter<User, UserCoord>::getRowIDsHostView(Zoltan2::BaseAdapter<User>::ConstIdsHostView &) const [with User=ztrowmatrix_t, UserCoord=ztrowmatrix_t]"
/home/ndellin/trilinos/Trilinos-pristine/packages/zoltan2/core/src/input/Zoltan2_TpetraRowMatrixAdapter.hpp(375): here
            implicit generation of "Zoltan2::TpetraRowMatrixAdapter<User, UserCoord>::~TpetraRowMatrixAdapter() noexcept [with User=ztrowmatrix_t, UserCoord=ztrowmatrix_t]"
/home/ndellin/trilinos/Trilinos-pristine/packages/zoltan2/core/src/input/Zoltan2_TpetraRowMatrixAdapter.hpp(375): here
            instantiation of class "Zoltan2::TpetraRowMatrixAdapter<User, UserCoord> [with User=ztrowmatrix_t, UserCoord=ztrowmatrix_t]"
/home/ndellin/trilinos/Trilinos-pristine/packages/zoltan2/core/src/input/Zoltan2_TpetraRowMatrixAdapter.hpp(375): here
            instantiation of "Zoltan2::TpetraRowMatrixAdapter<User, UserCoord>::TpetraRowMatrixAdapter(const Teuchos::RCP<const User> &, int) [with User=ztrowmatrix_t, UserCoord=ztrowmatrix_t]"
/home/ndellin/trilinos/Trilinos-pristine/packages/zoltan2/test/core/unit/input/TpetraRowMatrixInput.cpp(208): here

/home/ndellin/trilinos/Trilinos-pristine/packages/kokkos/core/src/impl/Kokkos_ViewMapping.hpp(3541): error: static assertion failed with "View assignment must have compatible spaces"
          detected during:
            instantiation of "void Kokkos::Impl::ViewMapping<DstTraits, SrcTraits, std::enable_if_t<<expression>, void>>::assign(Kokkos::Impl::ViewMapping<DstTraits, SrcTraits, std::enable_if_t<<expression>, void>>::DstType &, const Kokkos::Impl::ViewMapping<DstTraits, SrcTraits, std::enable_if_t<<expression>, void>>::SrcType &, const Kokkos::Impl::ViewMapping<DstTraits, SrcTraits, std::enable_if_t<<expression>, void>>::TrackType &) [with DstTraits=Kokkos::ViewTraits<std::remove_const_t<const Tpetra::Details::DefaultTypes::global_ordinal_type> *, std::conditional_t<true, Kokkos::Cuda::array_layout, Kokkos::Cuda::array_layout>, Kokkos::Device<Kokkos::HostSpace::execution_space, Kokkos::Cuda::memory_space>, std::conditional_t<false, void, Kokkos::Experimental::DefaultViewHooks>>, SrcTraits=Kokkos::ViewTraits<std::remove_const_t<const Tpetra::Details::DefaultTypes::global_ordinal_type> *, std::conditional_t<true, Kokkos::Cuda::array_layout, Kokkos::Cuda::array_layout>, Kokkos::Device<Kokkos::HostSpace::execution_space, Kokkos::HostSpace::memory_space>, std::conditional_t<false, void, Kokkos::Experimental::DefaultViewHooks>>]"
/home/ndellin/trilinos/Trilinos-pristine/packages/kokkos/core/src/Kokkos_View.hpp(1335): here
            instantiation of "std::enable_if_t<Kokkos::Impl::ViewMapping<Kokkos::View<DataType, Properties...>::traits, Kokkos::View<RT, RP...>::traits, Kokkos::ViewTraits<DataType, Properties...>::specialize>::is_assignable_data_type, Kokkos::View<DataType, Properties...>> &Kokkos::View<DataType, Properties...>::operator=(const Kokkos::View<RT, RP...> &) [with DataType=std::remove_const_t<const Tpetra::Details::DefaultTypes::global_ordinal_type> *, Properties=<std::conditional_t<true, Kokkos::Cuda::array_layout, Kokkos::Cuda::array_layout>, Kokkos::Device<Kokkos::HostSpace::execution_space, Kokkos::Cuda::memory_space>, std::conditional_t<false, void, Kokkos::Experimental::DefaultViewHooks>>, RT=std::remove_const_t<const Tpetra::Details::DefaultTypes::global_ordinal_type> *, RP=<std::conditional_t<true, Kokkos::Cuda::array_layout, Kokkos::Cuda::array_layout>, Kokkos::Device<Kokkos::HostSpace::execution_space, Kokkos::HostSpace::memory_space>, std::conditional_t<false, void, Kokkos::Experimental::DefaultViewHooks>>]"
/home/ndellin/trilinos/Trilinos-pristine/packages/zoltan2/core/src/input/Zoltan2_TpetraRowMatrixAdapter.hpp(169): here
            instantiation of "void Zoltan2::TpetraRowMatrixAdapter<User, UserCoord>::getRowIDsHostView(Zoltan2::BaseAdapter<User>::ConstIdsHostView &) const [with User=ztrowmatrix_t, UserCoord=ztrowmatrix_t]"
/home/ndellin/trilinos/Trilinos-pristine/packages/zoltan2/core/src/input/Zoltan2_TpetraRowMatrixAdapter.hpp(375): here
            implicit generation of "Zoltan2::TpetraRowMatrixAdapter<User, UserCoord>::~TpetraRowMatrixAdapter() noexcept [with User=ztrowmatrix_t, UserCoord=ztrowmatrix_t]"
/home/ndellin/trilinos/Trilinos-pristine/packages/zoltan2/core/src/input/Zoltan2_TpetraRowMatrixAdapter.hpp(375): here
            instantiation of class "Zoltan2::TpetraRowMatrixAdapter<User, UserCoord> [with User=ztrowmatrix_t, UserCoord=ztrowmatrix_t]"
/home/ndellin/trilinos/Trilinos-pristine/packages/zoltan2/core/src/input/Zoltan2_TpetraRowMatrixAdapter.hpp(375): here
            instantiation of "Zoltan2::TpetraRowMatrixAdapter<User, UserCoord>::TpetraRowMatrixAdapter(const Teuchos::RCP<const User> &, int) [with User=ztrowmatrix_t, UserCoord=ztrowmatrix_t]"
/home/ndellin/trilinos/Trilinos-pristine/packages/zoltan2/test/core/unit/input/TpetraRowMatrixInput.cpp(208): here

2 errors detected in the compilation of "/home/ndellin/trilinos/Trilinos-pristine/packages/zoltan2/test/core/unit/input/TpetraRowMatrixInput.cpp".
make[2]: *** [packages/zoltan2/test/core/unit/CMakeFiles/Zoltan2_TpetraRowMatrixInput.dir/build.make:75: packages/zoltan2/test/core/unit/CMakeFiles/Zoltan2_TpetraRowMatrixInput.dir/input/TpetraRowMatrixInput.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:10274: packages/zoltan2/test/core/unit/CMakeFiles/Zoltan2_TpetraRowMatrixInput.dir/all] Error 2
```

This may follow from fairly recent changes with #12083 


### Steps to Reproduce
1. SHA1:  d454829507ada18ba38608cd43d24b7ae937d2b7
1. Configure script:  For Weaver testbed, rhel8 queue
```bash
# Allocate node
bsub -Is -n 1 -q rhel8 -gpu "num=1" bash

# Environment prep
export ATDM_CONFIG_REGISTER_CUSTOM_CONFIG_DIR=${TRILINOS_DIR}/cmake/std/atdm/contributed/weaver
source ${TRILINOS_DIR}/cmake/std/atdm/load-env.sh weaver-cuda-11.2-opt
export OMPI_CXX="$KOKKOS_PATH/bin/nvcc_wrapper"

# Configure
cmake \
      -DCMAKE_CXX_FLAGS='-g' \
      -DCMAKE_CXX_STANDARD="17" \
      -DCMAKE_INSTALL_PREFIX=$PWD/install \
      -DTrilinos_CONFIGURE_OPTIONS_FILE:STRING=cmake/std/atdm/ATDMDevEnv.cmake \
      -DTrilinos_ENABLE_COMPLEX_DOUBLE=ON \
      -DTrilinos_ENABLE_TESTS=OFF \
      -DTrilinos_ENABLE_ALL_PACKAGES=OFF \
      -DTPL_ENABLE_CUSPARSE:BOOL=ON \
      -DFC_FN_UNDERSCORE=UNDER \
      \
      -D Trilinos_ENABLE_Kokkos=ON \
      -D Kokkos_ARCH_VOLTA70=ON \
      -D Kokkos_ARCH_POWER9=ON \
      -D Kokkos_ENABLE_CUDA=ON \
      -D Kokkos_ENABLE_CUDA_LAMBDA=ON \
      -D Kokkos_ENABLE_CUDA_UVM=ON \
      -D KokkosKernels_INST_MEMSPACE_CUDAUVMSPACE=ON \
      -D Trilinos_ENABLE_Zoltan2=ON \
      -D Zoltan2_ENABLE_TESTS=ON \
      \
$TRILINOS_DIR

# Build
make -j
```

--- a/packages/zoltan2/core/src/input/Zoltan2_Adapter.hpp
+++ b/packages/zoltan2/core/src/input/Zoltan2_Adapter.hpp
@@ -124,14 +124,14 @@ public:
   using ConstOffsetsDeviceView = Kokkos::View<const offset_t *, device_t>;
   using ConstOffsetsHostView = typename ConstOffsetsDeviceView::HostMirror;
 
-  using offsetsDeviceView = Kokkos::View<offset_t *, device_t>;
-  using offsetsHostView = typename offsetsDeviceView::HostMirror;
+  using OffsetsDeviceView = Kokkos::View<offset_t *, device_t>;
+  using OffsetsHostView = typename OffsetsDeviceView::HostMirror;
 
   using ConstScalarsDeviceView = Kokkos::View<const scalar_t *, device_t>;
   using ConstScalarsHostView = typename ConstScalarsDeviceView::HostMirror;
 
-  using scalarsDeviceView = Kokkos::View<scalar_t *, device_t>;
-  using scalarsHostView = typename scalarsDeviceView::HostMirror;
+  using ScalarsDeviceView = Kokkos::View<scalar_t *, device_t>;
+  using ScalarsHostView = typename ScalarsDeviceView::HostMirror;
 
   using ConstWeightsDeviceView1D = Kokkos::View<const scalar_t *, device_t>;
   using ConstWeightsHostView1D = typename ConstWeightsDeviceView1D::HostMirror;

--- a/packages/zoltan2/core/src/input/Zoltan2_TpetraRowMatrixAdapter.hpp
+++ b/packages/zoltan2/core/src/input/Zoltan2_TpetraRowMatrixAdapter.hpp
@@ -163,10 +163,12 @@ public:
 
   void getRowIDsHostView(
       typename BaseAdapter<User>::ConstIdsHostView &rowIds) const override {
+
     auto kRowIds = rowMap_->getMyGlobalIndices();
-    auto hostRowIds = Kokkos::create_mirror_view(kRowIds);
-    Kokkos::deep_copy(hostRowIds, kRowIds);
-    rowIds = hostRowIds;
+    auto tmpIds = typename BaseAdapter<User>::IdsHostView("rowIds", kRowIds.extent(0));
+
+    Kokkos::deep_copy(tmpIds, kRowIds);
+    rowIds = tmpIds;
   }
 
   void getRowIDsDeviceView(
@@ -256,6 +258,7 @@ public:
       typename BaseAdapter<User>::WeightsHostView &weights) const {
     auto hostWeight = Kokkos::create_mirror_view(kRowWeights_);
     Kokkos::deep_copy(hostWeight, kRowWeights_);
+
     weights = hostWeight;
   }
 
@@ -287,10 +290,10 @@ private:
   ArrayRCP<StridedData<lno_t, scalar_t>> rowWeights_;
 
   // New Kokkos Type
-  Kokkos::View<offset_t *, device_t> kOffset_;
-  Kokkos::View<gno_t *, device_t> kColumnIds_;
-  Kokkos::View<scalar_t *, device_t> kValues_;
-  Kokkos::View<scalar_t **, device_t> kRowWeights_;
+  typename BaseAdapter<User>::OffsetsDeviceView kOffset_;
+  typename BaseAdapter<User>::IdsDeviceView kColumnIds_;
+  typename BaseAdapter<User>::ScalarsDeviceView kValues_;
+  typename BaseAdapter<User>::WeightsDeviceView kRowWeights_;
   Kokkos::View<bool *, host_t> numNzWeight_;
 
   int nWeightsPerRow_;

